### DISCUSSION TITLE
providers info currently scraped by ani-cli

### DISCUSSION THREAD

**Discussion by CoolnsX** [pystardust/ani-cli#786] [General] [answered]
As you are well aware that we scrape multiple providers to increase availibility of episode... I m gonna discuss about all those provider so you can choose one according to your needs..

There are ~7~ 3 providers to choose from that are also implemented and travers in same order as in script... 
~- mp4upload (pass ```-r1``` argument for selecting it as first provider for scraping)~
~- doodstream (pass ```-r2``` argument for selecting it as first provider for scraping)~
~- streamlare (pass ```-r3``` argument for selecting it as first provider for scraping)~
~- okru (pass ```-r4``` argument for selecting it as first provider for scraping)~
- xstreamcdn (pass ```-r1``` argument for selecting it as first provider for scraping)
- animixplay direct links (pass ```-r2``` argument for selecting it as first provider for scraping)
- gogoanime direct links (pass ```-r3``` argument for selecting it as first provider for scraping)

So I will go through each and every provider..
~1. **Mp4upload**~
~- it is a high speed provider that store episode highest quality possible depending upon in which quality the owner uploaded the episode~
~- it's certificate gives problem that's why sometime it doesn't work, and gives problem on android side (termux) while streaming on mpv even with referrer.... downloading works though... and due to this reason it is not selected as default~
~- another reason for not selecting it as default cause it requires referrer which is problem for android side.. cause vlc doesn't support referrer~
~- though it's selected as default for downloading on all OS...~

~2. **Doodstream**~
~- it's a medium speed provider that also store episode highest quality possible depending upon in which quality the owner uploaded the episode~
~- it's a cloudfare enabled provider... so sometimes it doesn't give links due to it's inconsistently cloudfare bot trigger that differs from device to device... so it's possible that doodstream provider works on one device and doesn't works on another device even if they are on same network..~
~- its selected as default provider for streaming on all OS except android cause it requires referrer and you know the rest..~

~3. **streamlare**~
~- it's a medium speed provider that also store episode highest quality possible depending upon in which quality the owner uploaded the episode.. unlike doodstream it doesn't require referrer..~
~- it's selected as default for streaming on android side cause of above reason..~

~4. **okru**~
~- its a medium-highish speed provider that stores multiquality links ranges from 144p to 1080p~
~- it stores episode having high file size like 1.4 GB for 1080p link.. so it becomes the drawback as it's constantly buffering if u don't have high speed internet..~
~- it's selected as default provider for scraping multi-quality links..~

5. **xstreamcdn**
- its a high speed provider that stores multi-quality links ranges from 480p to 1080p
- it doesn't require referrer so I consider it as my favourite provider cause of above point..

6. **animixplay direct links**
- it's a very high speed provider that stores multi-quality links ranges from 360p to 1080p 
- it's stores gogoanime direct links for very old anime like cowboy bebop so sometimes it can be slow due to this reason.. so I recommend you to download it first then watch.. ~cause there is a custom code on downloading side especially implemented for this provider that uses parallel downloading to maximise the use of bandwidth..~ (removed this and switched back to downloading through ffmpeg cause many user complained about video issues such as not working on other player, artifacts, glitches etc)
- but it's fast if you use it for watching recent anime... it's so fast that it can take up all your bandwidth..
~- on dowloading side.. it uses same custom code as gogoanime to download episode as fast as possible..~(problem explained above)
- one major drawback is it stores episodes having high file size same as okru provider.. but its so fast that it compensiate for it.. so if you have unlimited high speed internet then you can use this..

7. **gogoanime direct links**
- it a medium-highish speed provider that stores multi-quality links ranges from 360p to 1080p
- it's slow for old anime, but for recent release it's fast... 
- if u want to watch really old anime using this provider then ~I recommend you to download episode and then watch.. cause there is a custom code on downloading side especially implemented for this provider that uses parallel downloading to maximise the use of bandwidth..~ (problem explained above)

**Comment by YusufKhan-gamedev** [comment]
-fx not -rx

**Reply by justchokingaround** [reply]
it changed in latest version

**Reply by YusufKhan-gamedev** [reply]
oops, yeah, realized this was 27 days ago :rofl:

**Comment by Quicksilver151** [comment]
does this list just keep decreasing?

**Reply by CoolnsX** [reply]
This is final.. only selected fast and reliable providers

**Reply by Quicksilver151** [reply]
also it doesnt seem like this list is accurate

**Reply by Quicksilver151** [reply]
i was able to use 5 of those providers
i pulled the latest version of ani-cli just seconds ago

**Reply by CoolnsX** [reply]
there is a PR already opened.. so i updated the list beforehand..

**Comment by Quicksilver151** [comment]
would it be possible to add the names of the providers with their numbers in ani-cli's -h help section
searching for this discussion every time is a pain
just printing the index of the provider with the name will do

**Reply by CoolnsX** [reply]
Will do in next PR...

**Comment by CoolnsX** [comment, marked as answer]
this needs to be updated to match with current version..
but due to some changes at source end it's getting delayed..
